npep/npep-136.md

```diff
@@ -0,0 +1,102 @@
+   This implies that you should be able to create a policy that says “allow X”, where if X wasn’t already denied 
+   (eg by BaselineAdminNetworkPolicy) then the “allow X” would just be a no-op (rather than causing other traffic to be denied like with NPv1).
+   - It should not have the automatic “connections from the pod’s node IP are always accepted” exception.
+   (but you shouldn’t have to manually add that rule either… we probably need to make probes work differently)
+   TODO: add link to the probes discussion
+   - It should not have tricky “arrays of objects where multiple fields can be set but none are required” 
+   such that adding or removing a single “-” drastically changes the meaning of the policy.
+   - “Present but empty” / “Present but zero length” should not mean something different from “not present” / nil. 
+   Likewise, the difference between an array of length 0 and an array of length 1 should be the same as the difference 
+   between an array of length 1 and an array of length 2.
+
+DeveloperNetworkPolicy should work well together with ANP and BANP as a part of the same API.
+
+## User-Stories/Use-Cases
+
+### User Story 1: Isolate namespace
+
+As a namespace owner, I want to make sure my pods can only access other pods from the same namespace and nothing else.
+I also want to restrict incoming connections from outside the namespace.
+
+### User Story 2: Allowlist egress external endpoints, deny ingress
+
+As a namespace owner, I want to allow pods in my namespace to only access specific external endpoints 
+that I define by CIDR and/or FQDN and deny everything else, including incoming connections.
+
+### User Story 3: Communication between namespaces
+
+As an owner of multiple namespaces, I want to allow and deny connections between namespaces.
+I have 3 namespaces: `db`, `backend`, `server`, and I want to define the following policies:
+1. namespace `db` ony allows incoming connection from the `backend` namespace, all outgoing connection are denied.
+2. namespace `backend` only allows incoming connections from the `server` namespace, and only allows outgoing connections
+to the `db` namespace.
+3. `server` namespace allows all incoming connections, and only allows outgoing connections to the `backend` namespace.
+
+### User Story 4: Communication within a namespace
+
+As a namespace owner, I have 3 pods in my namespace: `db`, `backend`, `server`. I want to define the same policy as in Story 3,
+but inside one namespace.
+
+### User Story 5: TODO: there are still more cases to cover
+
+### Less obvious User Stories (may be better postponed to the next iteration, after we agree on the most basic cases)
+- external ingress to the namespace (requires integration with the Ingress/Gateway API)
+- using nodes as peers (struggling to figure out the use case, may be something related to hostNetwork pods)
+- using services as peers (was never properly discussed before, but sounds like a valid case)
```

Comment:
Yeah, IMO ingress subjects and (in-cluster) egress peers should almost always be expressed as services rather than as pods.

The problem people always had with this is that the reverse (egress subjects / ingress peers as Services) mostly doesn't make sense, and people didn't like the asymmetry.

Reply:
we probably can consider having separate definitions for egress and ingress (we are already doing that for ANP peers, so we can also do that for subject)